src/main/scala/com/twitter/finagle/postgres/PostgresClientImpl.scala

```diff
@@ -188,6 +187,18 @@ class PostgresClientImpl(
    */
   override def isAvailable: Boolean = status == Status.Open
 
+  override def listen(channel: String): Future[Listener] = {
+    if(Listeners.alreadyListening(channel)) {
+      return Future(Listeners.addConsumer(channel))
+    }
+    query(s"LISTEN $channel").map(_ => Listeners.addConsumer(channel))
```

Comment:
I think handling `LISTEN` properly is still more complicated than this.

This will `LISTEN` on *one connection in the connection pool*, of which there are many. If that connection goes away (which can happen all the time during normal operation), we will silently no longer be listening anymore.

I think rather than being global, each connection needs to track its listeners. Then we either need to prevent connections with listeners from being downscaled away, or (more likely) ensure that when a listening connection closes, another connection starts listening.

It gets really complicated, which is why I've put off supporting it :)

Reply:
Thanks for the feedback, @jeremyrsmith. Maybe track the connection end events on the state machine and re-calculate/re-send listen commands? Or throw an error so the client can reconnect? I'm a little bit busy these days, but I'll put some thoughts on this ASAP.